[compiler-rt][AArch64] Provide basic implementations of SME memcpy/memmove in case of strictly aligned memory access

vhscampos opened the pull request:
The existing implementations, written in assembly, make use of unaligned accesses for performance reasons. They are not compatible with strict aligned configurations, i.e. with `-mno-unaligned-access`.

If the functions are used in this scenario, an exception is raised due to unaligned memory accesses.

This patch reintroduces vanilla implementations for these functions to be used in strictly aligned configurations. The actual code is largely based on the code from https://github.com/llvm/llvm-project/pull/77496

vhscampos committed:
[compiler-rt][AArch64] Provide basic implementations of SME memcpy/memmove in case of strictly aligned memory access

The existing implementations, written in assembly, make use of unaligned
accesses for performance reasons. They are not compatible with strict
aligned configurations, i.e. with `-mno-unaligned-access`.

If the functions are used in this scenario, an exception is raised due
to unaligned memory accesses.

This patch reintroduces vanilla implementations for these functions to
be used in strictly aligned configurations. The actual code is largely
based on the code from https://github.com/llvm/llvm-project/pull/77496

david-arm commented on compiler-rt/lib/builtins/aarch64/sme-libc-routines.c:
```diff
@@ -22,3 +23,46 @@ const void *__arm_sc_memchr(const void *src, int c,
+
+#ifndef __ARM_FEATURE_UNALIGNED
+
+static void *memcpy_fwd(void *dest, const void *src,
```
Is it still worth adding the `__arm_sc_` prefix to these functions to keep all functions in the same namespace? Just worried about potential clashes with other target implementations.

david-arm replied:
OK yeah I see. If it's ok with you, I do think it's better to have the same naming scheme for all the functions and start with `__arm_sc_`

vhscampos replied:
Done. Thanks

david-arm commented on compiler-rt/lib/builtins/aarch64/sme-libc-routines.c:
```diff
@@ -22,3 +23,46 @@ const void *__arm_sc_memchr(const void *src, int c,
+  const unsigned char *srcp = (const unsigned char *)src;
+
+  if ((srcp > (destp + n)) || (destp > (srcp + n)))
+    return __arm_sc_memcpy(dest, src, n);
```
I assume you're calling `__arm_sc_memcpy` because it has the `__restrict` attributes on the pointers and you're hoping the compiler will take advantage of that during compilation?

vhscampos replied:
Yes. From some experimentation, the compiler does take advantage of it: https://godbolt.org/z/oj4b97KK9

david-arm reviewed (approved):
LGTM!

sdesmalen-arm commented on compiler-rt/lib/builtins/aarch64/sme-libc-routines.c:
```diff
@@ -22,3 +23,46 @@ const void *__arm_sc_memchr(const void *src, int c,
   return NULL;
 }
+
+#ifndef __ARM_FEATURE_UNALIGNED
```
To avoid creating many different variants and various ifdefs, does it make sense to choose the implementation (either `sme-libc-mem-routines.S` (optimised) or `sme-libc-routines.c` (basic)) in CMakeLists.txt ? That way we can also let cmake emit a warning that the unoptimised SME routines are chosen, because of the constraints given by the compilation flags to compiler-rt.

vhscampos replied:
The issue is it isn't either this or that. For example, for AArch64 soft-nofp with unaligned accesses, some of the functions will come from the optimised .S file, and others from the basic .c file.

In order to have it work without the #ifdefs and avoid multiple definition errors, I will have to refactor the files into smaller ones and select them accordingly in the CMakeLists.txt. I am having a go at this option.

vhscampos committed:
Changes:
- Split functions into separate files.
- Select which implementation to use based on target features. The
  selection is now done in CMake.

vhscampos commented on compiler-rt/lib/builtins/CMakeLists.txt:
```diff
@@ -600,9 +600,17 @@ if (COMPILER_RT_HAS_AARCH64_SME)
     set_source_files_properties(aarch64/arm_apple_sme_abi.s PROPERTIES COMPILE_FLAGS -march=armv8a+sme)
     message(STATUS "AArch64 Apple SME ABI routines enabled")
   elseif (NOT COMPILER_RT_DISABLE_AARCH64_FMV AND COMPILER_RT_HAS_FNO_BUILTIN_FLAG AND COMPILER_RT_AARCH64_FMV_USES_GLOBAL_CONSTRUCTOR)
-    list(APPEND aarch64_SOURCES aarch64/sme-abi.S aarch64/sme-libc-mem-routines.S aarch64/sme-abi-assert.c aarch64/sme-libc-routines.c)
```
The change inadvertently removed files from the build. I will fix it shortly.
